### ISSUE TITLE
Missing prop-types validation for certain edge cases

### ISSUE THREAD

**Issue by ViggoV** [yannickcr/eslint-plugin-react#2126]
Prop-types validation seems to be broken in certain cases, but I have a hard time pinning down exactly when. I have seen it only in functional components wrapped in multiple Higher Order Components, specifically Redux' `connect()` as well as ReactDnD's `DropTarget()` and `DragSource()`.

**ENV**
node: v8.11.4
npm: 6.5.0
eslint: v5.12.0
eslint-plugin-react: 7.12.3

**Code**
Works: The following component behaves as expected, showing an error for the missing prop-types:
```javascript
import React from 'react';
import { connect } from 'react-redux';
import Header from './Header';
import Cell from './Cell';

const Layout = (props) => (
  <div className="layout-container">
    <Header />
    {
      props.layout.rows.map(r => 
        <div key={'row-'+r.id} className="layout-row" style={/*styles.row*/}>
          { r.cells.map(c => <Cell key={'cell-'+c.id} row_id={r.id} cell={c} />) }
        </div>
      )
    }
  </div>
);

const mapStateToProps = (state) => ({
 layout: state.layout 
});

export default connect(mapStateToProps)(Layout);
```

Doesn't work: This component shows no errors at all (but does show other linting errors)
```javascript
import React from  'react';
import { connect } from 'react-redux';
import { DragSource, DropTarget } from 'react-dnd';

import { 
  FiArrowRight, 
  FiArrowLeft,
  FiPlus,
  FiMinus,
  FiPlusSquare
} from 'react-icons/fi';

const Cell = props => {

  const { 
    cell, changeFlex, changeOrder,
    connectDragSource, connectDragPreview, isDragging,
    connectDropTarget,
    dragInProgress,
    newCellButton
  } = props;

  if (newCellButton) {
    return connectDragSource(<span><FiPlusSquare /></span>);
  }

  let cell_styles = {
    flex: dragInProgress ? 1 : cell.flex,
    order: cell.order,
    opacity: isDragging ? .5 : 1
  };

  return connectDragPreview(
    connectDropTarget(
      <div className="layout-cell" style={cell_styles}>
        <div className="padded-cell">
          { connectDragSource(
            <div className="cell-header">
              <div className="cell-title">
                {cell.title}
              </div>
              <div className="cell-toolbox">
                <div onClick={() => {changeFlex(-1);}}><FiMinus /></div>
                <div>{cell.flex}</div>
                <div onClick={() => {changeFlex(1);}}><FiPlus /></div>

                <div onClick={() => {changeOrder(-1);}}><FiArrowLeft /></div>
                <div>{cell.order}</div>
                <div onClick={() => {changeOrder(1);}}><FiArrowRight /></div>
              </div>
            </div>
          )}
          <div className="cell-contents">{cell.title}</div>
        </div>
      </div>
    )
  );
};

const mapStateToProps = (state) => ({
  dragInProgress: state.drag.dragInProgress
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  changeFlex: (flex_delta) => { 
    const { row_id, cell: { id: cell_id } } = ownProps;
    dispatch({ type: 'LAYOUT_CELL_CHANGE_FLEX', row_id, cell_id, flex_delta}); 
  },
  changeOrder: (order_delta) => { 
    const { row_id, cell: { id: cell_id } } = ownProps;
    dispatch({ type: 'LAYOUT_CELL_CHANGE_ORDER', row_id, cell_id, order_delta}); 
  },
  setDragState: (newState) => {
    if (newState) dispatch({type: 'DRAG_BEGIN'});
    else dispatch({type: 'DRAG_END'});
  },
  moveCell: (src, tgt) => {
    dispatch({
      type: 'LAYOUT_MOVE_CELL',
      src,
      tgt
    });
  }
});

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(DropTarget(
  'layout_cell',
  {
    hover(props,monitor) {
      let src = monitor.getItem(),
          tgt = { 
            id: props.cell.id, 
            row_id: props.row_id,
            order: props.cell.order
          };
      if (src.id != tgt.id) {
        props.moveCell(src, tgt);
        src.order = tgt.order;
        src.row_id = tgt.row_id;
      }
    }
  },
  (connect, monitor) => ({
    connectDropTarget: connect.dropTarget(),
    getItem: monitor.getItem()
  })
)(DragSource(
  'layout_cell',
  { 
    beginDrag(props) {
      props.setDragState(true); 
      return { 
        id: props.cell.id,
        row_id: props.row_id,
        order: props.cell.order
      }; 
    },
    endDrag(props) {
      props.setDragState(false);
    },
    isDragging(props,monitor) {
      return props.cell.id === monitor.getItem().id;
    } 
  },
  (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    connectDragPreview: connect.dragPreview(), 
    isDragging: monitor.isDragging()
  })
)(Cell)));
```
**Discrepancy between Editor and CLI** 
For whatever reason, and this is beyond me, SublimeLinter picks up the first example just fine and draws a squiggly line under the code:
![skaermbillede 2019-01-09 kl 10 01 10 3](https://user-images.githubusercontent.com/3153568/50888331-b9679780-13f5-11e9-9347-c60baa4ed0ec.png)

but running `$ eslint ./src` in CLI all I get is:
<img width="574" alt="skaermbillede 2019-01-09 kl 09 52 12" src="https://user-images.githubusercontent.com/3153568/50887865-912b6900-13f4-11e9-9956-5275109dcd17.png">

**eslint config**
I use a single .eslintrc.json in my home folder (`~/`, I'm on a Mac) looking like:
```json
{
    "env": {
        "browser": true,
        "commonjs": true,
        "es6": true,
        "node": true,
        "mocha": true
    },
    "extends": [
      "eslint:recommended",
      "plugin:react/recommended",
      "p5js",
      "p5js/dom"
    ],
    "parserOptions": {
        "sourceType": "module",
        "ecmaVersion": 2018,
        "ecmaFeatures": {
          "JSX": true
        }
    },
    "rules": {
        "indent": "off",
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "single"
        ],
        "semi": [
            "error",
            "always"
        ],
        "no-console": [
          "warn"
        ],
        "multiline-comment-style": [
          "warn",
          "separate-lines"
        ]
    },
    "plugins": [
      "react"
    ],
    "settings": {
      "react": {
        "version": "detect"
      }
    }
}
```

And thanks for a great plugin :)

**Comment by ViggoV** [comment]
Update: Turns out my SublimeLinter vs. CLI discrepancy was due to me missing the `--ext .jsx` flag in the CLI. The main issue, however, remains..

**Comment by ljharb** [comment]
In the examples, you’re not using typical HOC patterns - you’re returning a function call which receives a jsx element as an argument (instead of passing a full component into an HOC, and using the return result as another component). I don’t know how a linter could possibly realize that those function calls return jsx themselves, nor how functions returning non-jsx values could be detected as components.

**Comment by ViggoV** [comment]
Ah, so the issue is the connect functions around the jsx itself? That makes a lot of sense! I just got smarter, thank you 👍 

I'll a workaround.

**Comment by ljharb** [comment]
Happy to reopen if you’re still having trouble.

**Comment by ViggoV** [comment]
It worked smoothly! Thanks for the quick and friendly reply 👍
